### PR TITLE
feat: cleanup pod critical pod annotations feature

### LINKED ISSUE #79548: Remove Critical Pod Annotation
[Critical Pod Annotation](https://github.com/kubernetes/kubernetes/blob/665e76d97623447d13bb3b33b68591a985b49c9d/pkg/features/kube_features.go#L58) is an experimental feature that lets Kubernetes control plane know about a pod being a critical pod by adding a specific annotation to the pod. This was added in 1.5 when K8s didn't have pod priority. Pod priority allows us to mark pods as critical with our first-class API. As a result, this feature was deprecated in 1.13. We can now remove the feature from our code base.

/sig scheduling
/sig node

### PR DESCRIPTION
**Pull request by draveness**
/kind cleanup
/priority important-soon
/assign @bsalamat 
/sig scheduling
/sig node

**What this PR does / why we need it**:

> Critical Pod Annotation is an experimental feature that lets Kubernetes control plane know about a pod being a critical pod by adding a specific annotation to the pod. This was added in 1.5 when K8s didn't have pod priority. Pod priority allows us to mark pods as critical with our first-class API. As a result, this feature was deprecated in 1.13. We can now remove the feature from our code base.

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes/kubernetes/issues/79548

**Does this PR introduce a user-facing change?**:

```release-note
scheduler.alpha.kubernetes.io/critical-pod annotation is removed. Pod priority (spec.priorityClassName) should be used instead to mark pods as critical. Action required!
```

### CODE REVIEW

File: cluster/addons/calico-policy-controller/calico-node-daemonset.yaml [hunk 1]
```diff
@@ -18,7 +18,6 @@ spec:
       labels:
         k8s-app: calico-node
       annotations:
```
Comment (mattjmcnaughton) [inline cluster/addons/calico-policy-controller/calico-node-daemonset.yaml]: Should we delete `annotations` entirely?

File: cluster/addons/calico-policy-controller/calico-node-vertical-autoscaler-deployment.yaml [hunk 2]
```diff
@@ -17,7 +17,6 @@ spec:
       labels:
         k8s-app: calico-node-autoscaler
       annotations:
-        scheduler.alpha.kubernetes.io/critical-pod: ''
```
Comment (mattjmcnaughton) [inline cluster/addons/calico-policy-controller/calico-node-vertical-autoscaler-deployment.yaml]: Do we need to add anything to mark these pods as critical with the first-class api that replaced critical pod annotations?

**Review by mattjmcnaughton — COMMENTED** [review]
Two high level questions on the yaml from me - this code touches some parts I'm not super familiar with so going to leave the full review to someone else :)

File: cluster/addons/calico-policy-controller/calico-node-daemonset.yaml (hunk 1 above)
Comment (draveness) [reply cluster/addons/calico-policy-controller/calico-node-daemonset.yaml]: IMO this is not necessary, but if someone is not happy with this. I'd love to remove this from.

File: cluster/addons/calico-policy-controller/calico-node-vertical-autoscaler-deployment.yaml (hunk 2 above)
Comment (draveness) [reply cluster/addons/calico-policy-controller/calico-node-vertical-autoscaler-deployment.yaml]: That makes sense, but I need some suggestions from folks who are familiar with these files.

File: cluster/addons/calico-policy-controller/calico-node-daemonset.yaml (hunk 1 above)
Comment (draveness) [reply cluster/addons/calico-policy-controller/calico-node-daemonset.yaml]: I just took a look at all the yaml files, they all have `priorityClass: system-node-critical` to set them to critical.

File: cluster/addons/calico-policy-controller/calico-node-vertical-autoscaler-deployment.yaml (hunk 2 above)
Comment (draveness) [reply cluster/addons/calico-policy-controller/calico-node-vertical-autoscaler-deployment.yaml]: I just took a look at all the yaml files, they all have `priorityClass: system-node-critical` to set them to critical.

**Comment by draveness** [comment]
/retest

Regards,
Draven

File: cluster/addons/calico-policy-controller/calico-node-daemonset.yaml (hunk 1 above)
Comment (BenTheElder) [reply cluster/addons/calico-policy-controller/calico-node-daemonset.yaml]: I don't think we should have the key with no values but :man_shrugging:

**Review by BenTheElder — COMMENTED** [review]
are these consumed only by the current release?

File: pkg/controller/daemon/daemon_controller_test.go [hunk 3]
```diff
@@ -1778,126 +1776,6 @@ func setDaemonSetToleration(ds *apps.DaemonSet, tolerations []v1.Toleration) {
 	ds.Spec.Template.Spec.Tolerations = tolerations
 }
 
-// DaemonSet should launch a pod even when the node with MemoryPressure/DiskPressure/PIDPressure taints.
-func TestTaintPressureNodeDaemonLaunchesPod(t *testing.T) {
```
Comment (bsalamat) [inline pkg/controller/daemon/daemon_controller_test.go]: Why are we dropping these tests?

File: pkg/kubelet/eviction/eviction_manager_test.go [hunk 4]
```diff
@@ -1147,11 +1148,13 @@ func TestCriticalPodsAreNotEvicted(t *testing.T) {
 		podStats[pod] = podStat
 	}
 
-	// Mark the pod as critical
+	// Mark the pod as static
```
Comment (bsalamat) [inline pkg/kubelet/eviction/eviction_manager_test.go]: You can drop this comment now.

File: pkg/kubelet/eviction/eviction_manager_test.go [hunk 5]
```diff
@@ -1147,11 +1148,13 @@ func TestCriticalPodsAreNotEvicted(t *testing.T) {
...
 	pods[0].Annotations = map[string]string{
-		kubelettypes.CriticalPodAnnotationKey:  "",
 		kubelettypes.ConfigSourceAnnotationKey: kubelettypes.FileSource,
 	}
+	// Mark the pod as critical
+	clusterPriority := scheduling.SystemCriticalPriority
```
Comment (bsalamat) [inline pkg/kubelet/eviction/eviction_manager_test.go]: s/clusterPriority/podPriority/

File: pkg/kubelet/eviction/eviction_manager_test.go [hunk 6]
```diff
@@ -1252,24 +1252,6 @@ func TestCriticalPodsAreNotEvicted(t *testing.T) {
 	if manager.IsUnderMemoryPressure() {
 		t.Errorf("Manager should not report memory pressure")
 	}
-
-	// Disable critical pod annotation feature gate
-	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.ExperimentalCriticalPodAnnotation, false)()
-
-	// induce memory pressure!
```
Comment (bsalamat) [inline pkg/kubelet/eviction/eviction_manager_test.go]: Why are we dropping the lines below here?

File: pkg/kubelet/types/pod_update.go [hunk 7]
```diff
@@ -142,18 +138,13 @@ func (sp SyncPodType) String() string {
 	}
 }
 
-// IsCriticalPod returns true if the pod bears the critical pod annotation key or if pod's priority is greater than
-// or equal to SystemCriticalPriority. Both the default scheduler and the kubelet use this function
-// to make admission and scheduling decisions.
+// IsCriticalPod returns true if pod's priority is greater than or equal to SystemCriticalPriority.
+// Both the default scheduler and the kubelet use this function to make admission and scheduling
```
Comment (bsalamat) [inline pkg/kubelet/types/pod_update.go]: Please delete lines 142 and 143. The scheduler does not really care about criticality of pods. From the scheduler's point of view a pod with a higher priority is more important than pods with lower priority.

File: pkg/kubelet/types/pod_update_test.go [hunk 8]
```diff
@@ -118,7 +115,8 @@ func TestString(t *testing.T) {
 }
 
 func TestIsCriticalPod(t *testing.T) {
```
Comment (bsalamat) [inline pkg/kubelet/types/pod_update_test.go]: You can drop this test. The test below checks critical pods based on priority.

File: test/e2e_node/critical_pod_test.go [hunk 9]
```diff
@@ -1,153 +0,0 @@
...
-)
-
-const (
-	criticalPodName   = "critical-pod"
-	guaranteedPodName = "guaranteed"
-	burstablePodName  = "burstable"
-	bestEffortPodName = "best-effort"
-)
-
-var _ = framework.KubeDescribe("CriticalPod [Serial] [Disruptive] [NodeFeature:CriticalPod]", func() {
```
Comment (bsalamat) [inline test/e2e_node/critical_pod_test.go]: Do we have equivalent node tests that check priority-based critical pod eviction works? If not, we should keep these tests and change pod priorities to be considered critical.

**Review by bsalamat — COMMENTED** [review]
Thanks, @draveness!

**Comment by spiffxp** [comment]
/release-note-action-required
Since this is removing a deprecated feature, anyone using that deprecated feature needs to take action before upgrading

**Review by draveness — COMMENTED** [review]
Thanks @bsalamat, all the comments have been addressed, please take another look

File: pkg/controller/daemon/daemon_controller_test.go [hunk 10]
```diff
@@ -1812,90 +1812,12 @@ func TestTaintPressureNodeDaemonLaunchesPod(t *testing.T) {
 	}
 }
 
-// When ScheduleDaemonSetPods is disabled, DaemonSet should launch a critical pod even when the node has insufficient free resource.
-func TestInsufficientCapacityNodeDaemonLaunchesCriticalPod(t *testing.T) {
```
Comment (spiffxp) [inline pkg/controller/daemon/daemon_controller_test.go]: I still don't understand why these tests are being removed.  Is this functionality not applicable when using the equivalent PriorityClass?  Is this functionality proven by other tests instead?

**Review by spiffxp — COMMENTED** [review]
One question on _test changes, the cluster/ changes look good

**Comment by spiffxp** [comment]
/approve
the _test stuff is under pkg which I don't have rights for anyway, the test/ and cluster/ changes look good

File: pkg/controller/daemon/daemon_controller_test.go (hunk 10 above)
Comment (draveness) [reply pkg/controller/daemon/daemon_controller_test.go]: addressed

File: pkg/controller/daemon/daemon_controller_test.go [hunk 11]
```diff
@@ -1820,7 +1820,6 @@ func TestInsufficientCapacityNodeDaemonLaunchesCriticalPod(t *testing.T) {
 		ds := newDaemonSet("critical")
 		ds.Spec.UpdateStrategy = *strategy
 		ds.Spec.Template.Spec = podSpec
-		setDaemonSetCritical(ds)
```
Comment (bsalamat) [inline pkg/controller/daemon/daemon_controller_test.go]: If I understand correctly, this tests is about creating a critical daemon set. So, why not keeping this line to create a critical daemon set? In that case, I think the second for loop that you have added in this PR can be removed.

Comment (draveness) [reply pkg/controller/daemon/daemon_controller_test.go]: The first loop is for checking the daemon set pod was not scheduled and the second one checks critical daemonset. I could remove this if it is not necessary to check the result when daemonset is not critical.

Comment (bsalamat) [reply pkg/controller/daemon/daemon_controller_test.go]: Yeah, I thought it was not necessary to check that non-critical daemonsets are not scheduled, but let's leave the check in.

**Review by bsalamat — COMMENTED** [review]
/lgtm
/approve

Thanks, @draveness!

**Comment by draveness** [comment]
Kindly ping @derekwaynecarr for approval.

**Comment by liggitt** [comment]
/approve

update the release note with the specific annotation that is no longer honored (so people can easily search their manifests) and the specific field `spec.priorityClassName` and priority class they should use instead

**Comment by draveness** [comment]
PR has been rebased, please take another look, @bsalamat @liggitt, thank you!

**Comment by bsalamat** [comment]
We have to revert this PR, as it caused a regression for static pods: #80203. After the static pod issue is addressed, we can resubmit this PR.

**Comment by draveness** [comment]
> We have to revert this PR, as it caused a regression for static pods: #80203. After the static pod issue is addressed, we can resubmit this PR.

I created #80277 to revert this PR and reopen the issue

**Pull request merged**